scripts/fuse/elektra_fuse/mock_context.py

```diff
@@ -0,0 +1,72 @@
+import multiprocessing, os, pwd, re
+from collections import namedtuple
+from pathlib import Path
+
+ProcessContext = namedtuple('ProcessContext', ["ruid", "euid", "suid", "fuid", "rgid", "egid", "sgid", "fgid", "env", "argv", "cwd", "pw_dir", "pw_name", "pid"])
+
+def parse_process_context_of_pid(pid):
+    #parse user and group ids
+    with open("/proc/%d/status" % pid, "r") as f:
+        status = f.read()
+        ruid, euid, suid, fuid = [int(num) for num in re.findall("^Uid:\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s*$", status, re.MULTILINE)[0]]
+        rgid, egid, sgid, fgid = [int(num) for num in re.findall("^Gid:\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s*$", status, re.MULTILINE)[0]]
+
+    #parse environment variables
+    with open("/proc/%d/environ" % pid, "r") as f:
+        raw_env_str = f.read()
+        env = dict( key_value.split('=', 1) for key_value in raw_env_str.strip("\0").split("\0") )
+
+    #parse argv
+    with open("/proc/%d/cmdline" % pid, "r") as f:
+        cmdline = f.read()
+        argv = cmdline.strip("\0").split("\0")
+
+    #parse current working directory
+    cwd = Path("/proc/%d/cwd" % pid).resolve()
+
+    struct_passwd = pwd.getpwuid(ruid)
+    pw_dir = struct_passwd.pw_dir
+    pw_name = struct_passwd.pw_name
+    return ProcessContext(ruid, euid, suid, fuid, rgid, egid, sgid, fgid, env, argv, cwd, pw_dir, pw_name, pid)
+
+def _mock_process_context_and_run(process_context, func, args, kwargs):
+
+    #libelekra depends both on $HOME and the real user id (not the effective user id!) to resolve the user: namespace
+    #therefore, temporarely changing the effective user id does not work
+
+    #mock cwd
+    os.chdir(process_context.cwd)
+
+    #mock environment (minus $HOME)
```

Comment:
Should not be necessary, see #3651

Reply:
#3651 allows you to pass your own argv, envp. So you basically:

1. open /proc/pid/cmdline and store content to argv
3. open /proc/pid/environ and store content to envp
4. when doing kdb.goptsContract pass this argv&envp (instead of sys.argv, os.environ)

see also https://master.libelektra.org/src/bindings/swig/python/examples/example_gopts.py

Reply:
Thanks, but how to call the interface (or where to get the required arguments) was not of concern, but rather whether a call has any effects on future queries to the namespaces `user:` and `dir:`.

Reply:
I thought that the implementation is completely stateless, so that any query to FUSE can never influence another query? So I am still not sure what the question is.

These 3 statements should only influence the proc: and the cascading / namespace, and should have no effect on user: or dir: (nor system: nor proc:) namespaces. Does this answer your question?

Reply:
Take a look at `src/bindings/swig/python/examples/example_gopts.py`. Specifically this bit:

https://github.com/ElektraInitiative/libelektra/blob/5702ce0496823fdd12c4e1e6ff9d6d1ca1b71a40/src/bindings/swig/python/examples/example_gopts.py#L72-L77

You can pass any Python lists to `kdb.goptsContract`, it doesn't have to be `sys.argv` and `os.environ`. So in your case it would be something in `process_context`.

You should then pass the `contract` every time you call `kdb.KDB()`, e.g. here
https://github.com/ElektraInitiative/libelektra/blob/378c276ca07183e62180a72391eaaafa93b12c97/scripts/fuse/elektra_fuse/elektra_util.py#L37-L42

Strictly speaking, you should only need the `contract` when you want to access a `proc:/` or cascading key, but if you are unsure just use it everywhere.

---


> whether a call has any effects on future queries to the namespaces `user:` and `dir:`.

The `gopts` plugin only adds keys into the `proc:/` namespace. So using `goptsContract` does not have any influence on `user:/` or `dir:/` keys. It may, however, affect which keys are returned by a cascading lookup.

The `goptsContract` also only applies to the one instance that receives it during `kdbOpen()` (or `kdb.KDB()` in the Python Binding). Other instances that existed before or are created later on without the contract are not affected.

Reply:
Btw. and working directory is a completely different topic.

It works with `os.chdir(path)` (before opening `kdb`). It changes keys in `dir:` and cascading namespace.

Reply:
Thank you for the detailed elaboration! It turns out, my misunderstanding stemmed from me assuming all lines marked in green in the initial comment were of concern, whereas in reality only the last one was relevant.

The functionality is implemented as of a4f59b349a6eefd327facb1322b802da23fa2289.